### DISCUSSION TITLE
"init" method

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#7075] [Language Ideas] [answered]
### TLDR:
Required properties are awesome.
Constructors do assignments like required properties, except they can also run code.
A method should exist to allow code to automatically be run after a ```with``` or after required properties are provided.

### Suggestion: ```init()``` method
In the event that a class has a method named ```init()```, the compiler will automatically call this method after ``with``` or after required properties are provided.  init() may accept a single optional argument of the containing type.  If it does, it will be called with the original value of the object (if ```with``` was used) or ```null``` if construction was used.

ie:
```
public record Person {
   public required int Age {get; init;}

  protected virtual void init() {
    if(Age < 0){
      throw new ArgumentOutOfRangeException(nameof(Age));
    }
  }

  protected virtual void init(Person Original) {
    if(Age < 0){
      throw new ArgumentOutOfRangeException(nameof(Age));
    }
  }

}

...

var A = new Person(){ Age = 23 };
//A.init() is called.
var B = A with { Age = 24 };
//B.init(A) is called.
```

Today, there are three main types of "events" that objects go through:

* Contruction
* Destruction
* IDisposable.Dispose()

Require properties and records have

**Comment by svick** [comment]
This has been considered before, see for example the brief section about "Final initializers" in https://github.com/dotnet/csharplang/discussions/3707. Though I couldn't find a dedicated discussion for this feature.

I think the variant for `with` is supposed to be handled by custom `Clone` methods instead (see "Allowing users to define cloning" in https://github.com/dotnet/csharplang/discussions/3707), but that also currently does not exist in the language.

**Comment by TahirAhmadov** [comment, marked as answer]
https://github.com/dotnet/csharplang/discussions/6591
